src/llama-sampling.cpp

```diff
@@ -1398,7 +1401,13 @@ static void llama_sampler_penalties_accept(struct llama_sampler * smpl, llama_to
         return;
     }
 
+    if (ctx->prev.size() >= (size_t) ctx->penalty_last_n) {
+        if (--ctx->token_count[ctx->prev.front()] == 0) {
```

Comment:
I think an `assert` here would be worth it as a sanity check:

```suggestion
    if (ctx->prev.size() >= (size_t) ctx->penalty_last_n) {
        assert(ctx->token_count.at(ctx->prev.front()) > 0);
        if (--ctx->token_count[ctx->prev.front()] == 0) {
```